### PR TITLE
JDK-8305782: Provide MacroAssembler::breakpoint on aarch64

### PR DESCRIPTION
**Pull request by tstuefe**
The ability to emit debug traps was useful for me on arm, and I miss it on aarch64.

Tested manually on Linux aarch64 in gdb with various values for hint covering the whole 16-bit range set. Hint gets encoded in the instruction (gdb decodes instruction as "BRK xxx" with xxx being the hint). According to documentation the hint ends up in ESR.ELx.ISS after the trap hit, but gdb refused to display the ESR register, so I could not verify that.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305782](https://bugs.openjdk.org/browse/JDK-8305782): Provide MacroAssembler::breakpoint on aarch64

### Reviewers
 * [Eric Liu](https://openjdk.org/census#eliu) (@theRealELiu - Committer)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/13401/head:pull/13401` \
`$ git checkout pull/13401`

Update a local copy of the PR: \
`$ git checkout pull/13401` \
`$ git pull https://git.openjdk.org/jdk.git pull/13401/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 13401`

View PR using the GUI difftool: \
`$ git pr show -t 13401`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/13401.diff">https://git.openjdk.org/jdk/pull/13401.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/13401#issuecomment-1501196806)

### CODE REVIEW

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 1]
```diff
@@ -2520,6 +2520,12 @@ void MacroAssembler::resolve_global_jobject(Register value, Register tmp1, Regis
   bind(done);
 }
 
+void MacroAssembler::breakpoint(uint16_t hint_imm16) {
+  // BRK with hint
+  const uint32_t c = 0xD4200000 | (((uint32_t)hint_imm16) << 5);
```
Comment (feilongjiang) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:2525]: aarch64 already provides `brk` at assembler [1],  for readability, maybe we can just simply use `brk(hint_imm16)` here.

1. https://github.com/openjdk/jdk/blob/969a6b9fd7f7afc60250309f3ada205c1473cf8e/src/hotspot/cpu/aarch64/assembler_aarch64.hpp#L1036

**Comment by tstuefe** [comment]
Thanks @feilongjiang; I reuse brk now.

File: src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp [hunk 2]
```diff
@@ -140,7 +140,7 @@ address LIR_Assembler::int_constant(jlong n) {
   }
 }
 
-void LIR_Assembler::breakpoint() { Unimplemented(); }
+void LIR_Assembler::breakpoint() { __ breakpoint(); }
```
Comment (theRealAph) [inline src/hotspot/cpu/aarch64/c1_LIRAssembler_aarch64.cpp:143]:
```suggestion
void LIR_Assembler::breakpoint() { __ brk(0); }
```

I don't think we need the rest of this patch.

**Review by theRealELiu — APPROVED** [review]
LGTM.

**Comment by tstuefe** [comment]
Withdrawn since the usefulness is questioned (one can just use brk(0))

**Comment by theRealAph** [comment]
The best way is like this:

```
address poo;

void stuff() {
  ... instructions ...
  poo = pc();
  ... instructions ...
}
```

then, in gdb:

```
(gdb) b <somewhere after startup>
breakpoint 2 at 0xabcedf
(gdb) comm 2
Type commands for breakpoint(s) 2, one per line.
End with a line saying just "end".
>b *poo
>c 
>end
(gdb)
```

Now, every time you run the program, a breakpoint will be set at `*poo`. And you can continue after the breakpoint.

**Comment by fisk** [comment]
> The best way is like this:
> 
> 
> 
> ```
> 
> address poo;
> 
> 
> 
> void stuff() {
> 
>   ... instructions ...
> 
>   poo = pc();
> 
>   ... instructions ...
> 
> }
> 
> ```
> 
> 
> 
> then, in gdb:
> 
> 
> 
> ```
> 
> (gdb) b <somewhere after startup>
> 
> breakpoint 2 at 0xabcedf
> 
> (gdb) comm 2
> 
> Type commands for breakpoint(s) 2, one per line.
> 
> End with a line saying just "end".
> 
> >b *poo
> 
> >c 
> 
> >end
> 
> (gdb)
> 
> ```
> 
> 
> 
> Now, every time you run the program, a breakpoint will be set at `*poo`. And you can continue after the breakpoint.

That doesn't work with code relocation, does it?

**Comment by tstuefe** [comment]
> The best way is like this:
> 
> ```
> address poo;
> 
> void stuff() {
>   ... instructions ...
>   poo = pc();
>   ... instructions ...
> }
> ```
> 
> then, in gdb:
> 
> ```
> (gdb) b <somewhere after startup>
> breakpoint 2 at 0xabcedf
> (gdb) comm 2
> Type commands for breakpoint(s) 2, one per line.
> End with a line saying just "end".
> >b *poo
> >c 
> >end
> (gdb)
> ```
> 
> Now, every time you run the program, a breakpoint will be set at `*poo`. And you can continue after the breakpoint.

Thanks, Andrew, that is helpful. I use breakpoint mostly in cases where I have a utility function that gets emitted multiple times, and I don't care which instance I debug, I'm fine with the first one that hits break.

**Pull request closed without merging**